src/mlpack/methods/ann/layer/neural_turing_machine_impl.hpp

```diff
@@ -0,0 +1,249 @@
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_NTM_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_NTM_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "neural_turing_machine_impl.hpp"
+
+#include "../visitor/forward_visitor.hpp"
+#include "../visitor/forward_with_memory_visitor.hpp"
+#include "../visitor/backward_with_memory_visitor.hpp"
+#include "../visitor/backward_visitor.hpp"
+#include "../visitor/gradient_visitor.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template <typename InputDataType, typename OutputDataType>
+NeuralTuringMachine<InputDataType, OutputDataType>::NeuralTuringMachine(
+    const size_t inSize,
+    const size_t outSize,
+    const size_t numMem,
+    const size_t memSize,
+    const size_t shiftSize,
+    LayerTypes controller) :
+    inSize(inSize),
+    outSize(outSize),
+    numMem(numMem),
+    memSize(memSize),
+    shiftSize(shiftSize),
+    controller(controller),
+    deterministic(false)
+{
+  network.push_back(controller);
+
+  readMem = new ReadMemory<>(outSize, numMem, memSize, shiftSize);
+  writeMem = new WriteMemory<>(outSize, numMem, memSize, shiftSize);
+
+  network.push_back(readMem);
+  network.push_back(writeMem);
+
+  allZeros = arma::zeros<arma::mat>(memSize, 1);
+
+  allOnes = arma::ones<arma::mat>(numMem, memSize);
+
+  memoryHistory.push_back(std::move(arma::mat(allOnes.memptr(),
+      allOnes.n_rows, allOnes.n_cols, false, true)));
+
+  bMemoryHistory = memoryHistory.end();
+
+  lReads.push_back(std::move(arma::mat(allZeros.memptr(),
+      allZeros.n_rows, allZeros.n_cols, false, true)));
+
+  gReads = lReads.end();
+
+  forwardStep = 0;
+  backwardStep = 0;
+  gradientStep = 0;
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void NeuralTuringMachine<InputDataType, OutputDataType>::Forward(
+    arma::Mat<eT>&& input, arma::Mat<eT>&& output)
+{
+  // Create input to the controller.
+  arma::vec newInput = arma::join_vert(input, lReads.back());
+
+  // Forward pass through the controller.
+  boost::apply_visitor(ForwardVisitor(std::move(newInput), std::move(
+      boost::apply_visitor(outputParameterVisitor, controller))),
+      controller);
+
+  // Get controller output.
+  arma::mat& controllerOutput = boost::apply_visitor(outputParameterVisitor,
+      controller);
+
+  // Acess to current memory.
+  arma::mat& cMemory = memoryHistory.back();
+
+  // Pass the controller output through read memory layer.
+  boost::apply_visitor(ForwardWithMemoryVisitor(std::move(controllerOutput),
+      std::move(cMemory),
+      std::move(boost::apply_visitor(outputParameterVisitor, readMem))),
+      readMem);
+  lReads.push_back(boost::apply_visitor(outputParameterVisitor, readMem));
```

Comment:
So it looks like you have built a lot of code like `ForwardWithMemory()` and `BackwardWithMemory()` and the associated visitors so that you can pass an `arma::mat memory` to the read memory and write memory.  But I think you could remove the need for all that extra code if `ReadMemory` and `WriteMemory` also held a reference to `memoryHistory`.  You could modify the constructor of `ReadMemory`:

```
ReadMemory(const size_t, const size_t, const size_t, const size_t, std::list<arma::mat>& memoryHistory)
```

and then add a member `std::list<arma::mat>& memoryHistory` to `ReadMemory`.  (Or maybe that would need to be a pointer so that serialization and assignment wouldn't have a problem.)  And you could do the same thing for `WriteMemory`.

Alternately, and I think this may be the better idea, simply add `arma::mat* currentMemory` members to both `ReadMemory` and `WriteMemory`, as well as a `CurrentMemory()` function to access and modify the pointer.  Then, instead of calling `ForwardWithMemoryVisitor`... do this:

```
readMem->CurrentMemory() = &memoryHistory.back();
boost::apply_visitor(ForwardVisitor(std::move(controllerOutput),
    std::move(boost::apply_visitor(outputParameterVisitor, readMem))),
    readMem);
```

and then use the same idea for the other instances where you would otherwise use `ForwardWithMemoryVisitor` or `BackwardWithMemoryVisitor`.

The nice thing about this situation is that you always know that the only classes that need to hold memory during `Forward()`, `Backward()`, or `Gradient()` will only be a `ReadMemory` or `WriteMemory`, so you can simply write the functions you need to set the memory correctly and access those directly, instead of through the visitor pattern.  The visitor pattern I think is more necessary for the `FFN` class, which is different because the actual layers that make up the FFN are in general not known, so we must use boost::variant + the visitor paradigm.

Let me know what you think of that idea.